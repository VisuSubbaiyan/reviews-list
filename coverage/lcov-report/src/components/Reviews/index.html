<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/Reviews</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> src/components/Reviews
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">45% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>18/40</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">37.5% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>6/16</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.18% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>7/17</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">46.15% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>18/39</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="Reviews.js"><a href="Reviews.js.html">Reviews.js</a></td>
	<td data-value="40.54" class="pic low"><div class="chart"><div class="cover-fill" style="width: 40%;"></div><div class="cover-empty" style="width:60%;"></div></div></td>
	<td data-value="40.54" class="pct low">40.54%</td>
	<td data-value="37" class="abs low">15/37</td>
	<td data-value="37.5" class="pct low">37.5%</td>
	<td data-value="16" class="abs low">6/16</td>
	<td data-value="37.5" class="pct low">37.5%</td>
	<td data-value="16" class="abs low">6/16</td>
	<td data-value="41.67" class="pct low">41.67%</td>
	<td data-value="36" class="abs low">15/36</td>
	</tr>

<tr>
	<td class="file high" data-value="Reviews.styled.js"><a href="Reviews.styled.js.html">Reviews.styled.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon May 27 2019 19:38:35 GMT+0200 (CEST)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
