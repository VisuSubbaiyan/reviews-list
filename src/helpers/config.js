export const theme = {
  primary: '#20232a',
  secondary: '#ccc',
  black: '#282c34',
  mediumGrey: '#aaa',
  lightGrey: '#dcdcdc',
  mediumFont: '14px',
  smallFont: '10px',
  linkColor: '#0366d6',
  blue: '#39a2d6',
  blueGrey: '#7998a8',
  formControlRadius: '5px',
  boxRadius: '10px',
  buttonRadius: '15px',
  breakpoints: {
    small: 640,
    medium: 769
  }
};
