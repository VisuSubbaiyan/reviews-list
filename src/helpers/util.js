import { css } from 'styled-components';
import { format as formatDate, endOfWeek, startOfWeek, startOfMonth } from 'date-fns';

/** Styled compoenent uitls */
// Based on the number of codes we can move this to seperate file
export const styling = props => props.css && css`${props.css}`;

export const viewport = breakpoint => (...args) => css`
    @media (min-width: ${breakpoint}px) {
      ${css(...args)}
    }
  `;

export const iconSizes = {
  xs: '12px',
  s: '16px',
  m: '24px',
  l: '32px',
  xl: '38px',
};

export const getSize = props => iconSizes[props.size];


/** Reviews component utils */
export const formatTitle = (title, group) => {
  switch (group) {
    case 'week':
      return `${formatDate(title, 'DD.MM')} - ${formatDate(endOfWeek(title), 'DD.MM.YYYY')}`;
    case 'month':
      return formatDate(title, 'MMMM YYYY');
    default:
      return formatDate(title, 'DD MMMM YYYY');
  }
};

export const isTextMatch = (text, match) => {
  const textLowerCase = text.toLowerCase();
  const matchLowerCase = match.toLowerCase();

  return textLowerCase.includes(matchLowerCase);
};

export const isReviewMatch = (review, matchWith) => {
  const { title, content } = review;

  return matchWith && (isTextMatch(title, matchWith) || isTextMatch(content, matchWith));
};

export const groupByIterator = group => review => {
  // GroupBy
  // Day new Date();
  // Week startOfWeek(new Date());
  // Month startOfMonth(new Date());
  const reviewedDate = new Date(review.reviewCreated);

  switch (group) {
    case 'week': {
      return startOfWeek(reviewedDate);
    }
    case 'month': {
      return startOfMonth(reviewedDate);
    }
    default: {
      return reviewedDate;
    }
  }
}

export const sortByOptions = [{ value: 'desc', name: 'Most Recent' }, { value: 'asc', name: 'Older' }];

export const groupByOptions = [
  { value: 'day', name: 'Group by day' },
  { value: 'week', name: 'Group by week' },
  { value: 'month', name: 'Group by month' }
];

export const filterOptions = [
  { value: 5, name: '5 star only' },
  { value: 4, name: '4 star only' },
  { value: 3, name: '3 star only' },
  { value: 2, name: '2 star only' },
  { value: 1, name: '1 star only' }
];

export const smoothScrollToPosition = top => window.scrollTo({
    top,
    behavior: 'smooth',
  });
