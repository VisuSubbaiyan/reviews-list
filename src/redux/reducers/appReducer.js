import {
  UPDATE_LOADING,
  UPDATE_REVIEWS,
  RESET_REVIEWS
} from '../actions';

/**
 * `INITIAL_STATE` is application initial state
 */
export const INITIAL_STATE = {
  page: 1,
  loading: false,
  reviews: [],
  hasMore: false
};

/**
 * `appReducer` - application reducer
 * @param {Object} state - application state
 * @param {Object} action - action object
 */
export const appReducer = (state = INITIAL_STATE, action) => {
  const { type, payload } = action;

  switch (type) {
    case UPDATE_LOADING: {
      const { loading } = payload;

      return ({ ...state, loading });
    }

    case UPDATE_REVIEWS: {
      const { reviews, hasMore, page } = payload;
      return ({ ...state, hasMore, page, reviews: [...state.reviews, ...reviews] });
    }

    case RESET_REVIEWS: {
      return ({ ...state, reviews: [] });
    }

    default:
      return state;
  }
};
