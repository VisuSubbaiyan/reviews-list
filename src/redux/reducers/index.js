// if we have multiple reducers,
// we can use combineReducers to make it as single reducer
export * from './appReducer';
