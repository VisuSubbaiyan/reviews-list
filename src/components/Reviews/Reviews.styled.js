import styled from 'styled-components';
import { viewport } from '../../helpers/util';

export const ReviewsMainWrapper = styled.div`
  padding-top: 150px;
  width: 95%;
  margin: 0 auto;
  min-height: 100vh;
  ${props => viewport(props.theme.breakpoints.medium)`width: 720px;`}
`;


export const StyledButton = styled.div`
  height: 20px;
  width: 20px;
  background: red;
`;