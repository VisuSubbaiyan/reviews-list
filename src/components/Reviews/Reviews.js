import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect, batch } from 'react-redux';
import { groupBy, orderBy, debounce, delay, isEmpty } from 'lodash';

import { formatTitle, isReviewMatch, groupByIterator, smoothScrollToPosition } from '../../helpers/util';
import { loadReviews, resetReviews } from '../../redux/actions';

import { Spinner, GeneralContiner, StyledButton } from '../base';
import { ReviewsMainWrapper } from './Reviews.styled';
import ReviewGroup from '../ReviewGroup';
import ReviewFilter from '../ReviewFilter/ReviewFilter';


/**
 * `Reviews` component responsible to load and render respective reviews
 */
class Reviews extends Component {
  /**
   * @property {<Redux.Action>} props.loadReview - action that load reviews from API
   * @property {<Redux.Action>} props.resetReviews - action that clear the reviews
   * @property {boolean} props.hasMore - indicates thats need to load more reviews or not
   * @property {boolean} props.loading - indicates thats loading is in progress or not
   * @property {Array<Object>} props.reviews - list of reviews
   */
  static propTypes = {
    hasMore: PropTypes.bool.isRequired,
    loading: PropTypes.bool,
    reviews: PropTypes.arrayOf(PropTypes.object).isRequired,
    loadReviews: PropTypes.func.isRequired,
    resetReviews: PropTypes.func.isRequired,
  }

  // UI state maintained in component state
  state = {
    search: '',
    stars: 5,
    groupBy: 'day',
    sortBy: 'desc',
  }

  componentDidMount() {
    this.props.loadReviews(this.props.page, this.state.sortBy);
    window.addEventListener('scroll', this.handleScrolling);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScrolling);
  }

  handleScrolling = () => {
    const { loading, hasMore } = this.props;

    if (loading || !hasMore) {
      return;
    }

    const lastChild = document.querySelector('div.reviews > div.review:last-child') || {};
    const lastIssueOffset = (lastChild.offsetTop + lastChild.clientHeight) - 300;
    const windowOffset = window.pageYOffset + window.innerHeight;

    if ((windowOffset > lastIssueOffset) || isEmpty(lastChild)) {
      this.props.loadReviews(this.props.page + 1, this.state.sortBy);
    }
  }

  handleRefresh = () => {
    this.setState({ sortBy: 'desc' });

    debounce(() => {
      const firstPage = 1;

      batch(() => {
        this.props.resetReviews();
        this.props.loadReviews(firstPage, this.state.sortBy);
      });

      delay(this.handleScrolling, 500);
    }, 300)();
  }

  handleFilters = filter => {
    this.setState(filter);

    delay(() => {
      this.handleScrolling();
      smoothScrollToPosition(0);
    }, 200);
  }

  renderReviews = () => {
    const { stars, groupBy: group, sortBy, search } = this.state;

    const filteredReviews = this.props.reviews
      .filter(review => review.stars === stars)
      .filter(review => !search || isReviewMatch(review, search));

    const groupedReviews = groupBy(filteredReviews, groupByIterator(group));
    const groupTitles = orderBy(Object.keys(groupedReviews), groupTitle => new Date(groupTitle), sortBy);

    if (groupTitles.length === 0) {
      this.handleScrolling();
      return;
    }

    return groupTitles.map(groupTitle => (
      <ReviewGroup
        key={groupTitle}
        groupTitle={formatTitle(groupTitle, group)}
        reviews={orderBy(groupedReviews[groupTitle], ({ reviewCreated }) => new Date(reviewCreated), sortBy)}
      />
    ));
  }

  render() {
    return (
      <ReviewsMainWrapper className="reviews">
        <ReviewFilter
          onChange={this.handleFilters}
          {...this.state}
          diableSort={this.props.hasMore}
        />
        <GeneralContiner css={`min-height: 0; justify-content: flex-end;`}>
          <StyledButton
            onClick={this.handleRefresh}
            type="button"
            value="Refresh"
            css={`position: absolute; margin-top: 15px;`}
          >
            {'Refresh'}
          </StyledButton>
        </GeneralContiner>
        {this.renderReviews()}
        {this.props.loading && <GeneralContiner><Spinner /></GeneralContiner>}
      </ReviewsMainWrapper>
    )
  }
}

export default connect(state => state, { loadReviews, resetReviews })(Reviews);
