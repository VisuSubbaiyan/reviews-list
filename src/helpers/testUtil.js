import React from 'react';
import { mount, shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { prettyPrint } from 'html';

import { theme } from './config';
import { UPDATE_LOADING } from '../redux/actions';

export const createMountedComponent = ({ children, isShallow }) => {
  const component = (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  );
  const mountedComponent = isShallow ? shallow(component) : mount(component);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const sampleText = 'sample text';
export const mockFunction = jest.fn(value => value);

export const inputProps = {
  value: 'sample name',
  onChange: mockFunction,
  placeholder: 'sample placeholder'
};

export const sampleState = {
  search: '',
  stars: 5,
  groupBy: 'day',
  sortBy: 'desc',
  diableSort: false
};

export const updateLoadingAction = {
  type: UPDATE_LOADING,
  payload: { loading: true }
};

export const expectedLoadingState = {
  page: 1,
  loading: true,
  reviews: [],
  hasMore: false
}
