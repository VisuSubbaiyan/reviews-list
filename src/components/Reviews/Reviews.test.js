import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import 'jest-styled-components';

import Reviews from './Reviews';

import { createMountedComponent } from '../../helpers/testUtil';
import { appReducer, INITIAL_STATE } from '../../redux/reducers';

jest.mock('../../services/reviewServices');

const store = createStore(appReducer, INITIAL_STATE, applyMiddleware(thunk));

describe('<Reviews /> should', () => {
  let mountedReview;

  beforeEach(() => {
    mountedReview = createMountedComponent({
      children: (
        <Provider store={store}>
          <Reviews loading={false} />
        </Provider>
      )
    });
  });

  it('render correctly with list of reviews', (done) => {
    setTimeout(() => {
      expect(mountedReview.htmlRender()).toMatchSnapshot();
      expect(mountedReview.find('Reviews').length).toEqual(1);
      expect(mountedReview.find('ReviewFilter').length).toEqual(1);

      done();
    });
  });

  xit('update reviews, hasMore in props', (done) => {
    setTimeout(() => {
      const reviewsInstance = mountedReview.find('Reviews').instance();

      expect(reviewsInstance.props.hasMore).toBeTruthy();
      expect(reviewsInstance.props.reviews.length).toEqual(20);

      done();
    });
  });
});
