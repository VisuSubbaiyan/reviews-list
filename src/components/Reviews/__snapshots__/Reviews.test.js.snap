// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reviews /> should render correctly with list of reviews 1`] = `
"<div class=\\"reviews sc-VigVT caOjsR\\">
  <div class=\\"sc-gzVnrw sc-kGXeez kBZLLG\\">
    <div class=\\"sc-kgoBCf gIdQai\\">
      <input placeholder=\\"Search reviews\\" class=\\"sc-iwsKbI bwkJWA\\" value=\\"\\">
      <select class=\\"sc-gqjmRU cxVyKR\\">
        <option value=\\"day\\">Group by day</option>
        <option value=\\"week\\">Group by week</option>
        <option value=\\"month\\">Group by month</option>
      </select>
      <select class=\\"sc-gqjmRU cxVyKR\\">
        <option value=\\"5\\">5 star only</option>
        <option value=\\"4\\">4 star only</option>
        <option value=\\"3\\">3 star only</option>
        <option value=\\"2\\">2 star only</option>
        <option value=\\"1\\">1 star only</option>
      </select>
      <select class=\\"sc-gqjmRU cxVyKR\\">
        <option value=\\"desc\\">Most Recent</option>
        <option value=\\"asc\\">Older</option>
      </select>
    </div>
  </div>
  <div class=\\"sc-gzVnrw gQTOSD\\">
    <button type=\\"button\\" value=\\"Refresh\\" class=\\"sc-bZQynM gpFfZL\\">Refresh</button>
  </div>
  <div class=\\"sc-gzVnrw gGsKwo\\">
    <div class=\\"sc-dnqmqq brBvzj\\"></div>
  </div>
</div>"
`;
